lib/vfs/src/lib.rs

```diff
@@ -0,0 +1,443 @@
+    WouldBlock,
+    /// A call to write returned 0
+    #[error("write returned 0")]
+    WriteZero,
+    /// Directory not Empty
+    #[error("directory not empty")]
+    DirectoryNotEmpty,
+    /// Some other unhandled error. If you see this, it's probably a bug.
+    #[error("unknown error found")]
+    UnknownError,
+}
+
+impl From<io::Error> for FsError {
+    fn from(io_error: io::Error) -> Self {
+        match io_error.kind() {
+            io::ErrorKind::AddrInUse => FsError::AddressInUse,
+            io::ErrorKind::AddrNotAvailable => FsError::AddressNotAvailable,
+            io::ErrorKind::AlreadyExists => FsError::AlreadyExists,
+            io::ErrorKind::BrokenPipe => FsError::BrokenPipe,
+            io::ErrorKind::ConnectionAborted => FsError::ConnectionAborted,
+            io::ErrorKind::ConnectionRefused => FsError::ConnectionRefused,
+            io::ErrorKind::ConnectionReset => FsError::ConnectionReset,
+            io::ErrorKind::Interrupted => FsError::Interrupted,
+            io::ErrorKind::InvalidData => FsError::InvalidData,
+            io::ErrorKind::InvalidInput => FsError::InvalidInput,
+            io::ErrorKind::NotConnected => FsError::NotConnected,
+            io::ErrorKind::NotFound => FsError::EntityNotFound,
+            io::ErrorKind::PermissionDenied => FsError::PermissionDenied,
+            io::ErrorKind::TimedOut => FsError::TimedOut,
+            io::ErrorKind::UnexpectedEof => FsError::UnexpectedEof,
+            io::ErrorKind::WouldBlock => FsError::WouldBlock,
+            io::ErrorKind::WriteZero => FsError::WriteZero,
+            io::ErrorKind::Other => FsError::IOError,
+            // if the following triggers, a new error type was added to this non-exhaustive enum
+            _ => FsError::UnknownError,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct ReadDir {
+    // TODO: to do this properly we need some kind of callback to the core FS abstraction
+    data: Vec<DirEntry>,
+    index: usize,
+}
+
+impl ReadDir {
+    pub fn new(data: Vec<DirEntry>) -> Self {
+        Self { data, index: 0 }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DirEntry {
+    pub path: PathBuf,
+    // weird hack, to fix this we probably need an internal trait object or callbacks or something
+    pub metadata: Result<Metadata, FsError>,
+}
+
+impl DirEntry {
+    pub fn path(&self) -> PathBuf {
+        self.path.clone()
+    }
+
+    pub fn metadata(&self) -> Result<Metadata, FsError> {
+        self.metadata.clone()
+    }
+
+    pub fn file_type(&self) -> Result<FileType, FsError> {
+        let metadata = self.metadata.clone()?;
+        Ok(metadata.file_type())
+    }
+
+    pub fn file_name(&self) -> OsString {
+        self.path
+            .file_name()
+            .unwrap_or_else(|| self.path.as_os_str())
+            .to_owned()
+    }
+}
+
+#[derive(Clone, Debug)]
+// TODO: review this, proper solution would probably use a trait object internally
```

Comment:
I don't understand what's wrong with this code. Can you elaborate please?

Reply:
Oh I see, maybe we should implement `Metadata` as a trait, and not as a struct. And I think it's a good idea indeed!